### PR TITLE
Add authentication (username & password) to AddMongoDB

### PR DESCRIPTION
**Pull request by Crazyht**
Please ensure to include optional authentication parameters, specifically a username and password, when utilizing the AddMongoDB function. In situations where these parameters are not explicitly defined, the default username "admin" will be utilized, and a password will be automatically generated. #4348 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4370)

### CODE REVIEW

**Comment by Crazyht** [comment]
How I can solve this test failure, in connection string password should be data-encoded, so I can't directly check "{mongodb-password.value}". 
Have you already a mechanism somewhere I can inspire of ? The only possibility I see is to force password like "aspire123" in test suit. Do you think is a good way to handle it ?

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs [hunk 1]
```diff
@@ -15,22 +17,39 @@ public static class MongoDBBuilderExtensions
     // Internal port is always 27017.
     private const int DefaultContainerPort = 27017;
 
+    private const string UserEnvVarName = "MONGO_INITDB_ROOT_USERNAME";
+    private const string PasswordEnvVarName = "MONGO_INITDB_ROOT_PASSWORD";
+    //MONGO_INITDB_DATABASE ???
```
Comment (davidfowl) [inline src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:22]: Delete?
```suggestion
```

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs [hunk 2]
```diff
@@ -6,24 +6,53 @@ namespace Aspire.Hosting.ApplicationModel;
 /// <summary>
 /// A resource that represents a MongoDB container.
 /// </summary>
-/// <param name="name">The name of the resource.</param>
-public class MongoDBServerResource(string name) : ContainerResource(name), IResourceWithConnectionString
+public class MongoDBServerResource : ContainerResource, IResourceWithConnectionString
```
Comment (davidfowl) [inline src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:9]: This is a breaking API change, we need to preserve the existing overload that takes a name.

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs [hunk 3]
```diff
@@ -6,24 +6,53 @@ namespace Aspire.Hosting.ApplicationModel;
...
+            ReferenceExpression.Create($"{UserNameParameter}") :
+            ReferenceExpression.Create($"{DefaultUserName}");
+
     /// <summary>
     /// Gets the connection string for the MongoDB server.
     /// </summary>
     public ReferenceExpression ConnectionStringExpression =>
         ReferenceExpression.Create(
-            $"mongodb://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+            $"mongodb://{UserNameReference}:{Uri.EscapeDataString(PasswordParameter.Value)}@{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
```
Comment (davidfowl) [inline src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:55]: Remove the `Uri.EscapeDataString`. This needs to be solved here https://github.com/dotnet/aspire/issues/3117

File: tests/Aspire.Hosting.Tests/ManifestGenerationTests.cs [hunk 4]
```diff
@@ -634,10 +634,23 @@ public void VerifyTestProgramFullManifest()
...
+                  "inputs": {
+                    "value": {
+                      "type": "string"
+                    }
+                  }
+                },
                 "mongodb": {
                   "type": "container.v0",
-                  "connectionString": "mongodb://{mongodb.bindings.tcp.host}:{mongodb.bindings.tcp.port}",
+                  "connectionString": "mongodb://admin:aspire123@{mongodb.bindings.tcp.host}:{mongodb.bindings.tcp.port}",
```
Comment (mitchdenny) [inline tests/Aspire.Hosting.Tests/ManifestGenerationTests.cs:648]: This should be referencing the parameter ... we should never have a literal secret here.

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs [hunk 5]
```diff
@@ -15,22 +17,39 @@ public static class MongoDBBuilderExtensions
...
     /// Adds a MongoDB resource to the application model. A container is used for local development. This version the package defaults to the 7.0.8 tag of the mongo container image.
     /// </summary>
     /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
     /// <param name="name">The name of the resource. This name will be used as the connection string name when referenced in a dependency.</param>
     /// <param name="port">The host port for MongoDB.</param>
+    /// <param name="userName">A parameter that contains the MongoDb server user name, or <see langword="null"/> to use a default value.</param>
+    /// <param name="password">A parameter that contains the MongoDb server password, or <see langword="null"/> to use a generated password.</param>
     /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
-    public static IResourceBuilder<MongoDBServerResource> AddMongoDB(this IDistributedApplicationBuilder builder, string name, int? port = null)
+    public static IResourceBuilder<MongoDBServerResource> AddMongoDB(this IDistributedApplicationBuilder builder,
```
Comment (eerhardt) [inline src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:33]: This is also a binary breaking change.

File: tests/testproject/TestProject.AppHost/TestProgram.cs [hunk 6]
```diff
@@ -120,8 +121,9 @@ public class TestProgram : IDisposable
             }
             if (!resourcesToSkip.HasFlag(TestResourceNames.mongodb))
             {
+                var mongoPassword = builder.AddParameter("mongodb-password");
                 var mongoDbName = "mymongodb";
-                var mongodb = AppBuilder.AddMongoDB("mongodb")
+                var mongodb = AppBuilder.AddMongoDB("mongodb", password: mongoPassword)
```
Comment (eerhardt) [inline tests/testproject/TestProject.AppHost/TestProgram.cs:126]: Existing tests shouldn't be modified. They should stay working after your change.

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs (hunk 2 above)
Comment (Crazyht) [reply src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs]: Is it better to overload the constructor or add new parameters with default to null?

File: tests/testproject/TestProject.AppHost/aspire-manifest.json [hunk 7]
```diff
@@ -262,7 +262,7 @@
     },
     "mongodb": {
       "type": "container.v0",
-      "connectionString": "mongodb://{mongodb.bindings.tcp.host}:{mongodb.bindings.tcp.port}",
+      "connectionString": "mongodb://admin:aspire123@{mongodb.bindings.tcp.host}:{mongodb.bindings.tcp.port}",
```
[the remark this answers is not in the document]
Comment (Crazyht) [reply tests/testproject/TestProject.AppHost/aspire-manifest.json]: Resolved by removing the Uri.EscapeDataString

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs (hunk 5 above)
Comment (Crazyht) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:32]: Is it better to add a new overload with new parameters?

File: tests/testproject/TestProject.AppHost/TestProgram.cs (hunk 6 above)
Comment (Crazyht) [reply tests/testproject/TestProject.AppHost/TestProgram.cs]: Resolved by removing the Uri.EscapeDataString

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs [hunk 8]
```diff
@@ -6,53 +6,48 @@ namespace Aspire.Hosting.ApplicationModel;
 /// <summary>
 /// A resource that represents a MongoDB container.
 /// </summary>
-public class MongoDBServerResource : ContainerResource, IResourceWithConnectionString
+/// <param name="name">The name of the resource.</param>
+/// /// <param name="userNameParameter">A parameter that contains the MongoDb server user name, or <see langword="null"/> to use a default value.</param>
+/// <param name="passwordParameter">A parameter that contains the MongoDb server password.</param>
+public class MongoDBServerResource(string name, ParameterResource? userNameParameter = null, ParameterResource? passwordParameter = null) : ContainerResource(name), IResourceWithConnectionString
```
Comment (davidfowl) [inline src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:12]: Is this also a breaking change?

Comment (Crazyht) [reply src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:12]: So, must I create an overload with new parameters instead of named parameters with defaults?

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs (hunk 5 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:32]: I'm not sure if we can avoid the breaking change here. If we add another overload with optional arguments how would you disambiguate between it and the new overload with the optional password arguments?

Comment (Crazyht) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:32]: I agree; otherwise, maybe the best option will be to get back to the original implementation and add an extension method of IResourceBuilder<MongoDBServerResource> (like "WithAuthentication(user?, pwd?)" ) to enable authentication for the MongoDB container. With this approach, we don't have any break changes, and we activate authentication only when the method is called (even if I wonder if running db without authentication is part of good practice:) ). Do you think this approach is more aligned?

Comment (davidfowl) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:32]: I think we should switch mongodb to always have a password (runtime breaking change in the name of security). The key here is to avoid removing the original method that will result in a method missing exception for those that compiled against the old one. To add a new method, we need the old method to make all of the other arguments non-optional and a new method with all of the parameters.

```C#
public static IResourceBuilder< MongoDBServerResource> AddMongoDb(this IDistributedApplicationBuilder builder, string name, int? port);
public static IResourceBuilder< MongoDBServerResource> AddMongoDb(this IDistributedApplicationBuilder builder, string name, int? port = null, IResourceBuilder<ParameterResource>? userName = null, IResourceBuilder<ParameterResource>? password = null);
```

Comment (Crazyht) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:32]: @davidfowl I'm okay with switching MongoBd secure by default, but in this case, I can't be fully secure without breaking change on MongoDBServerResource. Otherwise, someone can use the primary constructor with name only and default to MongoDBServerResource without username and password, even if I think nobody tries to create a MongoDBServerResource instance by hand. Let me know if you prefer I secure MongoDBServerResource also or if I code the version without breaking changes.

**Comment by Crazyht** [comment]
May I put a specific comment on `var passwordParameter = password?.Resource ?? ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-password", special: false);` to specify the "special: false" should be removed/reviewed when #3117 is done?

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs (hunk 5 above)
Comment (davidfowl) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:44]: If you manually create the resource, yes you're right. We can obsolete the constructor with no arguments and remove it in a later release. These 2 methods will always set the password.

**Comment by Crazyht** [comment]
I need to mark the MongoDBServerResource(string name) constructor as obsolete. I'm not happy with the Connection string management (due to the database connection string needing to insert the DB name in the middle of the URL).

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs [hunk 9]
```diff
@@ -10,20 +10,95 @@ namespace Aspire.Hosting.ApplicationModel;
...
     /// </summary>
-    public ReferenceExpression ConnectionStringExpression =>
-        ReferenceExpression.Create(
-            $"mongodb://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+    public ReferenceExpression ConnectionStringExpression => ReferenceExpression.Create($"{ConnectionStringWithoutOptionsExpression}{ConnectionStringOptionsExpression}");
+
+    /// <summary>
+    /// Gets the connection string for the MongoDB server without options parameters.
+    /// </summary>
+    public ReferenceExpression ConnectionStringWithoutOptionsExpression
```
Comment (davidfowl) [inline src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:70]: Do these need to be public?

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs [hunk 10]
```diff
@@ -10,20 +10,95 @@ namespace Aspire.Hosting.ApplicationModel;
...
+            {
+                return ReferenceExpression.Create($"mongodb://{UserNameReference}:{PasswordParameter}@{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Gets the options parameters for connection string of the MongoDB server.
+    /// </summary>
+    public ReferenceExpression ConnectionStringOptionsExpression
```
Comment (davidfowl) [inline src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:88]: Does this need to be public?

File: tests/testproject/TestProject.AppHost/TestProgram.cs [hunk 11]
```diff
@@ -51,7 +51,7 @@ public class TestProgram : IDisposable
         builder.Configuration["DcpPublisher:DeleteResourcesOnShutdown"] = "true";
         builder.Configuration["DcpPublisher:ResourceNameSuffix"] = $"{Random.Shared.Next():x}";
         builder.Configuration["DcpPublisher:RandomizePorts"] = randomizePorts.ToString(CultureInfo.InvariantCulture);
-
+
```
Comment (davidfowl) [inline tests/testproject/TestProject.AppHost/TestProgram.cs:54]:
```suggestion
```

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs [hunk 12]
```diff
@@ -10,20 +10,95 @@ namespace Aspire.Hosting.ApplicationModel;
...
     private EndpointReference? _primaryEndpoint;
 
+    /// <summary>
+    /// Initialize a resource that represents a MongoDB container.
+    /// </summary>
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="userNameParameter">A parameter that contains the MongoDb server user name, or <see langword="null"/> to use a default value.</param>
+    /// <param name="passwordParameter">A parameter that contains the MongoDb server password.</param>
+    /// <param name="authenticationMechanism">A parameter that contains the MongoDb server authentication mechanism, or <see langword="null"/> to use a default value.</param>
+    public MongoDBServerResource(string name, ParameterResource? userNameParameter, ParameterResource? passwordParameter, ParameterResource? authenticationMechanism) : this(name)
```
Comment (eerhardt) [inline src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:26]:
```suggestion
    public MongoDBServerResource(string name, ParameterResource? userNameParameter, ParameterResource passwordParameter, ParameterResource? authenticationMechanism) : this(name)
```

The rest of the resources that take a password parameter aren't nullable.

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs [hunk 13]
```diff
@@ -15,22 +17,57 @@ public static class MongoDBBuilderExtensions
...
+    /// <param name="userName">A parameter that contains the MongoDb server user name, or <see langword="null"/> to use a default value.</param>
+    /// <param name="password">A parameter that contains the MongoDb server password, or <see langword="null"/> to use a generated password.</param>
+    /// <param name="authenticationMechanism">A parameter that contains the MongoDb server authentication mechanism, or <see langword="null"/> to use a default value.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<MongoDBServerResource> AddMongoDB(this IDistributedApplicationBuilder builder,
+        string name,
+        int? port = null,
+        IResourceBuilder<ParameterResource>? userName = null,
+        IResourceBuilder<ParameterResource>? password = null,
+        IResourceBuilder<ParameterResource>? authenticationMechanism = null)
```
Comment (eerhardt) [inline src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:50]: Do we need this authentication mechanism as a public option? I don't see it in other resources - ex. PostgreSQL.

File: src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs [hunk 14]
```diff
@@ -10,20 +10,95 @@ namespace Aspire.Hosting.ApplicationModel;
...
+    /// Gets the parameter that contains the MongoDb server password.
+    /// </summary>
+    public ParameterResource? PasswordParameter { get; }
+
+    /// <summary>
+    /// Gets the parameter that contains the MongoDb server authentication mechanism.
+    /// </summary>
+    public ParameterResource? AuthenticationMechanismParameter { get; }
+
+    internal ReferenceExpression AuthenticationMechanismReference =>
```
Comment (eerhardt) [inline src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs:48]:
```suggestion
    private ReferenceExpression AuthenticationMechanismReference =>
```

File: tests/Aspire.Hosting.Tests/MongoDB/AddMongoDBTests.cs [hunk 15]
```diff
@@ -2,10 +2,13 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Net.Sockets;
+
 using Aspire.Hosting.MongoDB;
 using Aspire.Hosting.Tests.Utils;
 using Aspire.Hosting.Utils;
+
 using Microsoft.Extensions.DependencyInjection;
+
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Tests/MongoDB/AddMongoDBTests.cs:11]:
```suggestion
using Aspire.Hosting.MongoDB;
using Aspire.Hosting.Tests.Utils;
using Aspire.Hosting.Utils;
using Microsoft.Extensions.DependencyInjection;
```

These newlines are unnecessary.

File: tests/testproject/TestProject.AppHost/TestProgram.cs [hunk 16]
```diff
@@ -51,7 +51,6 @@ public class TestProgram : IDisposable
         builder.Configuration["DcpPublisher:DeleteResourcesOnShutdown"] = "true";
         builder.Configuration["DcpPublisher:ResourceNameSuffix"] = $"{Random.Shared.Next():x}";
         builder.Configuration["DcpPublisher:RandomizePorts"] = randomizePorts.ToString(CultureInfo.InvariantCulture);
-
```
Comment (eerhardt) [inline tests/testproject/TestProject.AppHost/TestProgram.cs:54]: This can be reverted as it is unnecessary.

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs (hunk 13 above)
Comment (davidfowl) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:50]: We don't

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs [hunk 17]
```diff
@@ -15,22 +17,57 @@ public static class MongoDBBuilderExtensions
...
+        var passwordParameter = password?.Resource ?? ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-password", special:false);
+
+        var mongoDBContainer = new MongoDBServerResource(name, userName?.Resource, passwordParameter, authenticationMechanism?.Resource);
+
+        var resBuilder = builder
             .AddResource(mongoDBContainer)
             .WithEndpoint(port: port, targetPort: DefaultContainerPort, name: MongoDBServerResource.PrimaryEndpointName)
             .WithImage(MongoDBContainerImageTags.Image, MongoDBContainerImageTags.Tag)
             .WithImageRegistry(MongoDBContainerImageTags.Registry);
+        if (mongoDBContainer.PasswordParameter is not null)
```
[the remark this answers is not in the document]
Comment (Crazyht) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:61]: Not in a manual crafting scenario, as I only mark MongoDBServerResource ctor obsolete, so we can't completely avoid null for extension. This test should be removed only when the obsolete MongoDBServerResource constructor is also removed.

File: src/Aspire.Hosting.MongoDB/PublicAPI.Shipped.txt [hunk 18]
```diff
@@ -8,12 +8,16 @@ Aspire.Hosting.ApplicationModel.MongoDBServerResource
 Aspire.Hosting.ApplicationModel.MongoDBServerResource.ConnectionStringExpression.get -> Aspire.Hosting.ApplicationModel.ReferenceExpression!
 Aspire.Hosting.ApplicationModel.MongoDBServerResource.Databases.get -> System.Collections.Generic.IReadOnlyDictionary<string!, string!>!
 Aspire.Hosting.ApplicationModel.MongoDBServerResource.MongoDBServerResource(string! name) -> void
+Aspire.Hosting.ApplicationModel.MongoDBServerResource.MongoDBServerResource(string! name, Aspire.Hosting.ApplicationModel.ParameterResource? userNameParameter, Aspire.Hosting.ApplicationModel.ParameterResource? passwordParameter) -> void
```
Comment (eerhardt) [inline src/Aspire.Hosting.MongoDB/PublicAPI.Shipped.txt:11]: We haven't shipped these APIs yet, so they should go in the Unshipped.txt file.

File: src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs:59]: > Not in a manual crafting scenario

This isn't the manual crafting scenario. We directly create the object on line 52 above and we always pass in a non-null `passwordParameter`.

File: tests/testproject/TestProject.AppHost/TestProgram.cs (hunk 16 above)
Comment (eerhardt) [inline tests/testproject/TestProject.AppHost/TestProgram.cs:54]: Can you revert this change so it doesn't go into source control history?

**Comment by davidfowl** [comment]
@eerhardt this should follow the persistence pattern we just added

**Comment by eerhardt** [comment]
> @eerhardt this should follow the persistence pattern we just added

It does already.

**Comment by davidfowl** [comment]
I think we need to test a deployment to ACA as well to make sure it still works.

**Comment by eerhardt** [comment]
Thanks @Crazyht for your contribution. I'd like to get this change in for .NET Aspire 9.0, but I'm unable to push to this branch because it is from a protected branch (your `main` branch).

So instead I created https://github.com/dotnet/aspire/pull/5788, which takes these changes, resolves the conflicts with the current code, and fixes the tests. Please take a look over there.

Closing in favor of https://github.com/dotnet/aspire/pull/5788.
